File: Sdks/DotNet/Metis.ClientSdk/Counter/AtomicCounter.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Metis.ClientSdk.Counter
{
    internal class AtomicCounter
    {
        private static AtomicCounter instance;

        private ConcurrentDictionary<String, Counter32> mapCounter32;
        private ConcurrentDictionary<String, Counter64> mapCounter64;

        public static AtomicCounter Instance
        {
            get
            {
                if (instance == null)
                {
                    instance = new AtomicCounter();
                }
                return instance;
            }
        }

        private AtomicCounter() 
        {
            //获取当前环境中处理器的数量
            int numProcs = Environment.ProcessorCount;
            //按照处理器数量的2倍, 预估对列表的同时操作线程数量
            mapCounter32 = new ConcurrentDictionary<string, Counter32>(numProcs * 2, 100);
            mapCounter64 = new ConcurrentDictionary<string, Counter64>(numProcs * 2, 100);
        }
        public int CounterCount
        {
            get { return mapCounter32.Count + mapCounter64.Count; }
        }
        public void Increase32(string key)
        {
            var counter = mapCounter32.GetOrAdd(key, (k) => new Counter32(k));
            counter.Increment();
        }
        public void Decrease32(string key)
        {
            var counter = mapCounter32.GetOrAdd(key, (k) => new Counter32(k));
            counter.Decrement();
        }
        public void Increase32(string key, Int32 num)
        {
            var counter = mapCounter32.GetOrAdd(key, (k) => new Counter32(k));
            counter.Increment(num);
        }
        public void Decrease32(string key, Int32 num)
        {
            var counter = mapCounter32.GetOrAdd(key, (k) => new Counter32(k));
            counter.Decrement(num);
        }
        public void Increase64(string key)
        {
            var counter = mapCounter64.GetOrAdd(key, (k) => new Counter64(k));
            counter.Increment();
        }
        public void Decrease64(string key)
        {
            var counter = mapCounter64.GetOrAdd(key, (k) => new Counter64(k));
            counter.Decrement();
        }
        public void Increase64(string key, Int64 num)
        {
            var counter = mapCounter64.GetOrAdd(key, (k) => new Counter64(k));
            counter.Increment(num);
        }
        public void Decrease64(string key, Int64 num)
        {
            var counter = mapCounter64.GetOrAdd(key, (k) => new Counter64(k));
            counter.Decrement(num);
        }
        public Int32 Reset32(string key)
        {
            var counter = mapCounter32.GetOrAdd(key, (k) => new Counter32(k));
            return counter.Reset();
        }
        public Int64 Reset64(string key)
        {
            var counter = mapCounter64.GetOrAdd(key, (k) => new Counter64(k));
            return counter.Reset();
        }
        public Counter32 Get32(string key)
        {
            Counter32 counter;
            if (!mapCounter32.TryGetValue(key, out counter))
            {
                return null;
            }
            return counter;
        }
        public Counter64 Get64(string key)
        {
            Counter64 counter;
            if (!mapCounter64.TryGetValue(key, out counter))
            {
                return null;
            }
            return counter;
        }
        public IList<KeyValuePair<string, Int64>> GetAll()
        {
            IList<KeyValuePair<string, Int64>> datas = new List<KeyValuePair<string, Int64>>();
            datas.AddRange(mapCounter32.ToArray().Select(o => new KeyValuePair<string, Int64>(o.Key, o.Value.Counter)));
            datas.AddRange(mapCounter64.ToArray().Select(o=>new KeyValuePair<string,Int64>(o.Key, o.Value.Counter)));
            return datas;
        }
        public void Clear()
        {
            mapCounter32.Clear();
            mapCounter64.Clear();
        }
    }
}

File: Sdks/Java/common/src/main/java/metis/client/sdk/counter/AtomicCounter.java
package metis.client.sdk.counter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Created by Administrator on 14-8-29.
 */
public class AtomicCounter {

    private static volatile AtomicCounter instance;

    private ConcurrentHashMap<String, Counter32> mapCounter32;
    private ConcurrentHashMap<String, Counter64> mapCounter64;

    public static AtomicCounter getInstance() {
        synchronized (AtomicCounter.class) {
            if(instance == null){
                instance = new AtomicCounter();
            }
            return instance;
        }
    }

    private AtomicCounter() {
        //获取当前环境的处理器数量
        int numProcs = Runtime.getRuntime().availableProcessors();
        //初始化队列,默认使用两倍的处理器数量
        mapCounter32 = new ConcurrentHashMap<String, Counter32>(numProcs*2);
        mapCounter64 = new ConcurrentHashMap<String, Counter64>(numProcs*2);
    }

    public int increase32(String key) {
       Counter32 counter = mapCounter32.putIfAbsent(key, new Counter32(key));
       return counter.increment();
    }

    public int  increase32(String key, String group) {
        Counter32 counter = mapCounter32.putIfAbsent(key, new Counter32(key, group));
        return counter.increment();
    }

    public long increase64(String key) {
        Counter64 counter = mapCounter64.putIfAbsent(key, new Counter64(key));
        return counter.increment();
    }

    public long  increase64(String key, String group) {
        Counter64 counter = mapCounter64.putIfAbsent(key, new Counter64(key, group));
        return counter.increment();
    }

    public int decrease32(String key) {
        if(!mapCounter32.containsKey(key))
            return -1;
        return mapCounter32.get(key).decrement();
    }

    public long decrease64(String key) {
        if(!mapCounter64.containsKey(key))
            return -1;
        return mapCounter64.get(key).decrement();
    }

    public void reset32(String key) {
        if(!mapCounter32.containsKey(key))
            return;
        mapCounter32.get(key).reset();
    }

    public void reset64(String key) {
        if(!mapCounter64.containsKey(key))
            return;
        mapCounter64.get(key).reset();
    }

    public void getAll() {
        HashMap<String, HashMap<String, Integer>> datas =
                new HashMap<String, HashMap<String, Integer>>();
    }

    public int counterCount() {
        return mapCounter32.size() + mapCounter64.size();
    }
}
